logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/incremental/IncrementalSyncManager.kt

```diff
@@ -61,45 +61,51 @@ internal class IncrementalSyncManager(
     @OptIn(ExperimentalCoroutinesApi::class)
     private val eventProcessingDispatcher = kaliumDispatcher.default.limitedParallelism(1)
 
-    private val coroutineExceptionHandler = CoroutineExceptionHandler { _, throwable ->
-        when (throwable) {
-            is CancellationException -> {
-                kaliumLogger.withFeatureId(SYNC).i("Sync job was cancelled")
-                incrementalSyncRepository.updateIncrementalSyncState(IncrementalSyncStatus.Pending)
-            }
-
-            is KaliumSyncException -> {
-                kaliumLogger.withFeatureId(SYNC).i("SyncException during events processing", throwable)
-                incrementalSyncRepository.updateIncrementalSyncState(IncrementalSyncStatus.Failed(throwable.coreFailureCause))
-            }
-
-            else -> {
-                kaliumLogger.withFeatureId(SYNC).i("Sync job failed due to unknown reason", throwable)
-                incrementalSyncRepository.updateIncrementalSyncState(IncrementalSyncStatus.Failed(CoreFailure.Unknown(throwable)))
-            }
-            // TODO: Trigger retry depending on failure
+    private val coroutineExceptionHandler = SyncExceptionHandler({
+        incrementalSyncRepository.updateIncrementalSyncState(IncrementalSyncStatus.Pending)
+    }, {
+        incrementalSyncRepository.updateIncrementalSyncState(IncrementalSyncStatus.Failed(it))
+        syncScope.launch {
+            delay(RETRY_DELAY)
+            startMonitoringForSync()
         }
-    }
+    })
 
     private val syncScope = CoroutineScope(SupervisorJob() + eventProcessingDispatcher)
 
     init {
+        startMonitoringForSync()
+    }
+
+    private fun startMonitoringForSync() {
         syncScope.launch(coroutineExceptionHandler) {
-            // TODO: Trigger re-sync when policy changes from DISCONNECT to KEEP_ALIVE
             slowSyncRepository.slowSyncStatus.collectLatest { status ->
                 if (status is SlowSyncStatus.Complete) {
-                    // START SYNC
-                    kaliumLogger.i("Starting QuickSync, as SlowSync is completed")
-                    doIncrementalSync()
+                    // START SYNC. The ConnectionPolicy doesn't matter the first time
+                    kaliumLogger.i("Starting IncrementalSync, as SlowSync is completed")
+                    doIncrementalSyncWhilePolicyAllows()
+                    incrementalSyncRepository.updateIncrementalSyncState(IncrementalSyncStatus.Pending)
```

Comment:
Is this needed? we are also calling to `updateIncrementalSyncState` after this enclosing `if`

Reply:
Yep.

While SlowSync is completed, IncrementalSync will:

Perform Sync, without any restriction on ConnectionPolicy.
If it ever ends, is because the policy is DISCONNECT_AFTER_PENDING_EVENTS, so we update status to Pending and start observing for policy upgrade.